provider/etcd/etcd_client.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	// create the configuration
+	etcdConfig := clientv3.Config{
+		Endpoints:        endpoints,
+		AutoSyncInterval: 60 * time.Second,
+		DialTimeout:      10 * time.Second,
+		DialOptions:      []grpc.DialOption{grpc.WithBlock()}, // block until connect
+	}
+
+	// create the client
+	client, err := clientv3.New(etcdConfig)
+	if err != nil {
+		// On startup, if we cannot connect to the etcd cluster, then fail hard so that the
+		// user may address a potential configuration issue.  Once the clientv3 connects
+		// successfully, clientv3 will reconnect to the etcd cluster as it goes down or up,
+		// or into or out of network connectivity.
+		level.Error(ec.logger).Log("msg", "Etcd connection failed", "err", err)
+		os.Exit(1)
+	} else {
+		level.Info(ec.logger).Log("msg", "Etcd connection successful")
+	}
+	ec.mtx.Lock()
+	ec.client = client
+	ec.mtx.Unlock()
+
+	// start a goroutine to ensure the client will be cleaned up when the context is done
+	go func() {
+		defer func() {
+			ec.mtx.Lock()
+			if ec.client != nil {
+				ec.client.Close()
+				ec.client = nil
+			}
+			ec.mtx.Unlock()
+			level.Info(ec.logger).Log("msg", "Etcd connection shut down")
+		}()
+
+		for range ctx.Done() {
+		}
+	}()
+	return ec, nil
+}
+
+func (ec *EtcdClient) CheckAndPut(alert *types.Alert) error {
+	// Reduce writes to Etcd.  Only put to Etcd if the current alert is different than the same
+	// alert in etcd (excluding/ignoring the alert.UpdatedAt field).  To do this we:
+	//   1) Fetch the alert with the same fingerprint from Etcd
+	//   2) Compare our current alert with the alert from Etcd.
+	//   3) If the alerts are different, then write it back.
+	//
+	// This mechanism is simple and convergent. It is more effiencient to get from Etcd, check,
+	// then put into Etcd because unnecessarily putting any alert into etcd will result in the
+	// put alert being sent to all the AMs which are watching etcd.
+
+	etcdAlert, err := ec.Get(alert.Fingerprint())
+	if err == nil {
+		if AlertsEqualExceptForUpdatedAt(etcdAlert, alert) {
+			return nil // skip write to etcd
+		}
+	}
+
+	return ec.Put(alert)
+}
+
+func (ec *EtcdClient) Get(fp model.Fingerprint) (*types.Alert, error) {
+	// We do a best effort.  If etcd is not initialized yet, then skip
+	if ec.client == nil {
+		level.Error(ec.logger).Log("msg", "Not getting alert from etcd, etcd not initialized yet")
+		return nil, ErrorEtcdNotInitialized
+	}
+
+	// ensure the operation does not take too long
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+
+	ec.mtx.Lock()
+	resp, err := ec.client.Get(ctx, ec.prefix+fp.String())
+	ec.mtx.Unlock()
+	if err != nil {
+		level.Error(ec.logger).Log("msg", "Error getting alert from etcd", "err", err)
+		return nil, err
+	}
+
+	if len(resp.Kvs) != 1 {
+		return nil, errors.New("etcdGet did not receive exactly one result for fingerprint")
+	}
+
+	alert, err := UnmarshallAlert(string(resp.Kvs[0].Value))
```

Comment:
SHOULD: Use one `l` not two. ref: https://golang.org/pkg/encoding/json/#Unmarshal

Reply:
Fixed